sycl/source/detail/program_manager/program_manager.cpp

```diff
@@ -2555,7 +2554,7 @@ device_image_plain ProgramManager::getDeviceImageFromBinaryImage(
     RTDeviceBinaryImage *BinImage, const context &Ctx, const device &Dev) {
   const bundle_state ImgState = getBinImageState(BinImage);
 
-  assert(compatibleWithDevice(BinImage, Dev));
+  assert(compatibleWithDevice(BinImage, *getSyclObjImpl(Dev).get()));
```

Comment:
Likewise, line 2554 would be a better change, IMO.

Reply:
What about L2567 where we are creating vector of devices? I thoughts it also should be refactored, but in a separate PR?

Reply:
What I've been doing during the transition is something like `std::vector<device>{createSyclObjFromImpl<device>(Dev.shared_from_this())}` but I can follow up on that separately.